### PR TITLE
Soloseng/publishing-version-fix

### PR DESCRIPTION
**Pull request by soloseng**
### Description

Updates the devchain anvil script to override the default version that is hardcoded into the package.json file

### CODE REVIEW

File: packages/protocol/scripts/change-anvil-devchain-package-version.ts
Comment (arthurgousset) [inline packages/protocol/scripts/change-anvil-devchain-package-version.ts:20]: Nit: This line prints a message, but I couldn't quite work out if and where the cleaning up happens. 
Is this only intended to print a message or am I missing something?

If it's only a message, is this necessary?

File: .github/workflows/protocol-devchain-anvil.yml
Comment (arthurgousset) [inline .github/workflows/protocol-devchain-anvil.yml]: Nit: Is this comment still needed? Is a safety check needed and/or performed anywhere?

**Review by arthurgousset — COMMENTED** [review]
(for future reference) Keeping track of context that helped review this PR.

 This is a follow-on PR for: 
- https://github.com/celo-org/celo-monorepo/pull/11041

This PR intends to fix a bug in the "Protocol Devchain Anvil / Generate anvil" workflow, in the "Publish @celo/devchain-anvil" step:

> npm error 403 403 Forbidden - PUT https://registry.npmjs.org/@celo%2fdevchain-anvil - You cannot publish over the previously published versions: 1.0.0.

Source: [CI](https://github.com/celo-org/celo-monorepo/actions/runs/9776057223/job/26987797774#step:24:44)

It seems like the `version` in `package.json` is not incremented correctly.

From [Slack](https://clabsco.slack.com/archives/C02LN9FRME1/p1719944737654929):
>  just notice that there was a small issue with the devchain publishing following #11041, where it kept trying to publish using the same version.
> I created [this fix](https://github.com/celo-org/celo-monorepo/pull/11105) for it, that essentially overrides the hardcoded version.
> Note that the `workflow_dispatch` wont work until this is merged to `master` branch.

**Review by arthurgousset — APPROVED** [review]
### Review

To be transparent, I'm a little confused about how this workflow works, so I'm not comfortable approving with high confidence. But, since this is a relatively low risk area of our codebase, because it only affects the anvil devchain package, I'm ok with "merging and seeing if it works". 

In the previous PR (#11041), you mentioned that you added a "workflow dispatch" to test this. As a comment below, could you ELI5 how that workflow can be used to test that the workflow works as intended? Also curious if you have a hunch why it didn't catch the bug fixed here.

### Other feedback

As a reviewer, it would have helped to have a slightly more verbose PR description. I went through [Slack](https://clabsco.slack.com/archives/C02LN9FRME1/p1719944737654929), this PR, and the [previous PR](https://github.com/celo-org/celo-monorepo/pull/11041) to piece together:
1. how things are intended to work,
2. where the bug is, and
3. where the bug is patched.

To be transparent, I still think I wouldn't be able to explain how this workflow works to a new-joiner without spending more time looking into it.

As feedback, part of the complexity (IMO) comes from the fact that we are using helper functions in `utils.ts` and other `.ts` scripts that are relatively shallow. That means, I have to step through various files to work out how something works, and we don't gain a ton from having small self-contained functions.

**Comment by soloseng** [comment]
Turns out the workflow dispatch only become available once merged to the master branch. I'll create a pr with this updated workflow on master shortly.

step-by-step of what this does:

1. creates an anvil devchain and outputs the final state of the devchain as `devchain.json`
2. Runs integration test to make sure everything is setup as expected
3. Determines the next release version based on the current branch or github tag.
    - Pre/post audit github tags -> sets NPM version with a pre/post audit tag
    - github release tag 'core-contracts.v${N}` -> sets the version latest stable version.
    - Release branch -> sets to canary version. 
    - Otherwise, does not increment the version, as it will only do a dry run
4. Updates the package.json file with the latest version
5. publishes the package to NPM, unless its a dry run.

**Pull request merged**
